proofs/tree.go

```diff
@@ -273,14 +276,16 @@ func NewDocumentTree(proofOpts TreeOptions) DocumentTree {
 		readablePropertyLengthSuffix = proofOpts.ReadablePropertyLengthSuffix
 	}
 	return DocumentTree{
-		propertyList:      []Property{},
-		merkleTree:        merkle.NewTreeWithOpts(opts),
-		salts:             salts,
+		propertyList:                 []Property{},
+		merkleTree:                   merkle.NewTreeWithOpts(opts),
+		salts:                        salts,
 		readablePropertyLengthSuffix: readablePropertyLengthSuffix,
-		leaves:            []LeafNode{},
-		hash:              proofOpts.Hash,
-		parentPrefix:      proofOpts.ParentPrefix,
-		compactProperties: proofOpts.CompactProperties,
+		leaves:                       []LeafNode{},
+		hash:                         proofOpts.Hash,
+		parentPrefix:                 proofOpts.ParentPrefix,
+		compactProperties:            proofOpts.CompactProperties,
+		leafNodesByReadableName:      make(map[string]LeafNode),
+		leafNodesByCompactsString:    make(map[string]LeafNode),
```

Comment:
lets not store the leafNode for the index. We just need to check if the leaf is already present. I would suggest the following

```
nameIndex, propertyIndex := make(map[string]struct{}), make(map[string]struct{})
```

The `struct{}` will have zero new allocations.

The usage would be:
```
_, ok := nameIndex[name] 
if !ok {
    return "duplicate"
}

nameIndex[name] = struct{}{}
```

Reply:
`leafNodesByReadableName` -> `nameIndex`
`leafNodesByCompactsString` -> `compactIndex`